docs/design/2022-04-21-performance-diagnosis-enhancements.md

```diff
@@ -0,0 +1,188 @@
+- Author(s): [cfzjywxk](https://github.com/cfzjywxk), [you06](https://github.com/you06)
+- Discussion PR:
+- Tracking Issue: #33963, #34106
+
+## Table of Contents
+
+* [Introduction](#introduction)
+* [Motivation or Background](#motivation-or-background)
+* [Detailed Design](#detailed-design)
+* [Impacts & Risks](#impacts--risks)
+
+## Introduction
+
+To enhance the diagnosis and tuning of TiDB, this RPC proposes some optimization of metrics, including the supplement of metrics, better display in Grafana.
+
+## Motivation or Background
+
+TiDB collects many metrics, but few of them are frequently used and the metrics of the critical paths are in different panels, which makes the diagnosis a problem. Furthermore, some latency events are not recorded in critical paths.
+
+This RPC proposes some improvements to the metrics collection, and display.
+
+## Detailed Design
+
+### Differences Between Internal and External Requests
+
+The tracking issue of it is #33963, and it should be applied to most of the metrics.
+
+#### Internal Transaction Usage
+
+The main goal of this task is to differentiate the internal and external transactions and mark the snapshots and transactions as internal for internal sessions and tasks.
+
+Some tasks are not running in the internal sessions, however, they are executed directly by `RunInNewTxn` :
+
+- DDL Backfill
+- Auto-ID management
+- Version check when bootstrapping
+- GC placement and label rules
+- Load data for cached tables
+- Alloc global ID for some operations
+    - Write sequence update to binlog
+    - Create temporary table
+
+When TiDB bootstrapping, it’ll create some internal sessions, these sessions are used for:
```

Comment:
Here we should take care of the two points bellow:
(1) The internal sessions created when TiDB bootstrapping may not be used to execute the internal transactions which could be executed on the session created by session pool.
(2) Currently, there is a bug when an internal session runs a transaction that it divides into several different transactions although it intends to execute in one transaction.

Reply:
> (1) ...

Sorry, I don't get the point, can you give an example?

> (2) ...

Do you mean that the internal transactions are not guaranteed to be atomic? It does not affect the resource statistic and diagnosis IMO.

Reply:
(1) for example, '(do *Domain) UpdateTableStatsLoop' has an input pram 'ctx sessionctx.Context' which is a session create ed in bootstrapping, but It uses 'se, err := h.pool.Get()'  to get a new session when executes some sqls. I see you intend to use InRestrictedSQL as the internal flag which has no problem to this.
(2) `Do you mean that some internal transactions are not guaranteed to be atomic?` 
Yes, here is a bug. And it uses different sessions in one transaction(actually, it's not one transaction)

Reply:
(2) is out of the goal of this PR, but I think it would not affect the aggregation result of resource statistics.